libshortfin/src/shortfin/local_system.h

```diff
@@ -0,0 +1,235 @@
+  }
+
+  // Initialization APIs. Calls to these methods is only permitted between
+  // construction and Initialize().
+  // ------------------------------------------------------------------------ //
+  void InitializeNodes(int node_count);
+  void InitializeHalDriver(std::string_view moniker,
+                           iree_hal_driver_ptr driver);
+  void InitializeHalDevice(std::unique_ptr<LocalDevice> device);
+  void FinishInitialization();
+
+ private:
+  void AssertNotInitialized() {
+    if (initialized_) {
+      throw std::logic_error(
+          "LocalSystem::Initialize* methods can only be called during "
+          "initialization");
+    }
+  }
+
+  const iree_allocator_t host_allocator_;
+
+  // NUMA nodes relevant to this system.
+  std::vector<LocalNode> nodes_;
+
+  // Map of retained hal drivers. These will be released as one of the
+  // last steps of destruction. There are some ancillary uses for drivers
+  // after initialization, but mainly this is for keeping them alive.
+  std::unordered_map<std::string_view, iree_hal_driver_ptr> hal_drivers_;
+
+  // Map of device name to a LocalSystemDevice.
+  std::vector<std::unique_ptr<LocalDevice>> retained_devices_;
+  std::unordered_map<std::string_view, LocalDevice *> named_devices_;
+  std::vector<LocalDevice *> devices_;
+
+  // VM management.
+  iree_vm_instance_ptr vm_instance_;
+
+  // Workers.
+  std::vector<std::unique_ptr<Worker>> workers_;
+
+  // Whether initialization is complete. If true, various low level
+  // mutations are disallowed.
+  bool initialized_ = false;
+};
+using LocalSystemPtr = std::shared_ptr<LocalSystem>;
+
+// Base class for configuration objects for setting up a LocalSystem.
+class SHORTFIN_API LocalSystemBuilder {
+ public:
+  LocalSystemBuilder(iree_allocator_t host_allocator)
+      : host_allocator_(host_allocator) {}
+  LocalSystemBuilder() : LocalSystemBuilder(iree_allocator_system()) {}
+  virtual ~LocalSystemBuilder() = default;
+
+  iree_allocator_t host_allocator() { return host_allocator_; }
+
+  // Construct a LocalSystem
+  virtual LocalSystemPtr CreateLocalSystem() = 0;
+
+ private:
+  const iree_allocator_t host_allocator_;
+};
+
+// Each device exists in the local system as part of some topology that consists
+// of the following levels:
+//
+//   Level 0: User device category / system driver prefix
+//     (i.e. "hip", "cuda", "local").
+//   Level 1: Device instance ordinal.
+//   Level 2: Instance topology vector representing the logical organization
+//     of the queues on the device instance.
+//
+// Concretely, this means that each leaf LocalSystemDevice instance consists
+// of an iree_hal_device_t (as managed by an iree_hal_driver_t) and a
+// single bit position within an iree_hal_queue_affinity_t. The total number
+// of devices of a class is thus equal to the product of the device instance
+// ordinal and every entry of the instance topology vector. There can be at
+// most 64 queues on a device instance.
```

Comment:
does ordinal not mean something roughly equivalent to device index? If so would the product be of the length of the set of ordinals and every instance of the topology vector?

Reply:
Just convention... we tend to call things "ordinals" when they were numbered specifically that way, presumably for reasons. Whereas "index" just happens to be whatever the enumeration order was. It means something specific on many of these machines as there is a concept of a physical position on the bus (regardless of whether devices are masked/not enumerated), and that maps to other physical resources (like numa nodes, memory spaces, etc).

With all of that said, I used it imprecisely here. Updated to:

// The total number
// of devices of a class is thus equal to the product of the device instance
// arity and the arity of the topology vector. There can be at most 64 queues 
// on a device instance.